lib-js/spec-directory/virtual-directory.ts

```diff
@@ -0,0 +1,203 @@
+import fs from "fs"
+import path from "path"
+import { Readable } from "stream"
+import SpecDirectory, { SpecIteratee } from "./spec-directory"
+import { archiveFromStream, Directory as HrxDirectory } from "node-hrx"
+import { RunOptions } from "../options"
+import { toHrx } from "./hrx"
+import { withAsyncCleanup } from "./cleanup"
+
+function createFileCache(dir: HrxDirectory) {
+  const cache: Record<string, string> = {}
+  for (const itemName of dir) {
+    const subitem = dir.get(itemName)
+    if (subitem?.isFile()) {
+      cache[itemName] = subitem.body
+    }
+  }
+  return cache
+}
+
+function createSubdirCache(dir: HrxDirectory) {
+  const cache: Record<string, HrxDirectory> = {}
+  for (const itemName of dir) {
+    const subitem = dir.get(itemName)
+    if (subitem?.isDirectory()) {
+      cache[itemName] = subitem
+    }
+  }
+  return cache
+}
```

Comment:
The file cache is needed because the node-hrx `Directory` doesn't allow updating file contents. I opted to not store the original `Directory` object because, once the directory object is created, it can be modified and the original contents of the `HrxDirectory` will be outdated. It's a way to safeguard against future contributors, say, trying to implement another file listing method by calling `HrxDirectory.list()`, since that will be outdated and incorrect if there have been changes to the filesystem.

Reply:
Gotcha. Ideally we'd update `node-hrx` to support mutating files, but that may not be worth it in the short term. Can you document this in the code?